contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -163,27 +149,74 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
+    address[] memory targetsTopUp = new address[](targetAddresses.length);
+    for (uint256 idx = 0; idx < targetAddresses.length; idx++) {
+      address targetAddress = targetAddresses[idx];
+      target = s_targets[targetAddress];
+      (bool needsFunding, ) = _needsFunding(targetAddress, target.minBalance);
+      if (
+        target.topUpAmount > 0 && localBalance >= target.topUpAmount && !(targetAddress == address(0)) && needsFunding
+      ) {
+        targetsTopUp[idx] = targetAddress;
+        localBalance -= target.topUpAmount;
+      } else {
+        emit TopUpBlocked(targetAddress);
       }
-      (bool needsFunding, address target) = _needsFunding(targetAddresses[idx], minBalance);
-      if (!needsFunding) {
-        emit TopUpBlocked(targetAddresses[idx]);
+    }
+    for (uint256 idx = 0; idx < targetsTopUp.length; idx++) {
+      address targetAddress = targetsTopUp[idx];
+      if (targetAddress == address(0)) {
         continue;
       }
-      LINK_TOKEN.transfer(target, topUpAmount);
-      emit TopUpSucceeded(targetAddresses[idx]);
+      target = s_targets[targetAddress];
+      bool success = i_linkToken.transfer(targetAddress, target.topUpAmount);
+      if (success) {
+        target.lastTopUpTimestamp = uint56(block.timestamp);
+        emit TopUpSucceeded(targetAddress);
+      } else {
+        emit TopUpFailed(targetAddress);
+      }
+    }
+  }
+
+  /// @notice checks the target (could be direct target or IAggregatorProxy), and determines
+  /// if it is elligible for funding
+  /// @param targetAddress the target to check
+  /// @param minBalance minimum balance required for the target
+  /// @return bool whether the target needs funding or not
+  /// @return address the address of the contract needing funding
+  function _needsFunding(address targetAddress, uint256 minBalance) private view returns (bool, address) {
+    uint256 minWaitPeriodSeconds = s_minWaitPeriodSeconds;
+    MonitoredAddress memory addressToCheck = s_targets[targetAddress];
+    ILinkAvailable target;
+    IAggregatorProxy proxy = IAggregatorProxy(targetAddress);
+    // Explicitly check if the targetAddress is the zero address
+    // or if it's not a contract. In both cases return with false,
+    // to prevent target.linkAvailableForPayment from running,
+    // which would revert the operation.
+    if (targetAddress == address(0) || !(targetAddress.code.length > 0)) {
+      return (false, address(0));
+    }
+    try proxy.aggregator() returns (address aggregatorAddress) {
+      if (aggregatorAddress == address(0)) {
+        return (false, address(0));
+      }
+      target = ILinkAvailable(aggregatorAddress);
+    } catch {
+      target = ILinkAvailable(targetAddress);
     }
+    try target.linkAvailableForPayment() returns (int256 balance) {
+      if (
+        balance < 0 ||
+        (uint256(balance) < minBalance && addressToCheck.lastTopUpTimestamp + minWaitPeriodSeconds <= block.timestamp)
```

Comment:
is this correct? should we skip the wait period check if balance <0? also I think that since minBalance is a uint96, it won't overflow if you convert it to an int256 (double check me here). So assuming we want the wait period to apply for all cases, we should be able to re-write this as...

```sol
if (balance < int256(minBalance) && addressToCheck.lastTopUpTimestamp + minWaitPeriodSeconds <= block.timestamp) {...}
```

Reply:
100% my fault, I missed this one. Changing it.